break down DistSender.Send

tbg opened the pull request:
- remove the mechanism for removing a cached leader upon
  caching the same leader twice in a row: silly idea, obviously
  concurrent request going to the wrong leader will return
  almost simultaneously and expire the freshly updated cache.
- the leaderCache now only deals with proto.Replica{}, not pointers.
- pretty complete refactoring of DistSender.Send to get more manageable
  pieces and less error handling spagetthi.
- atomic updates in ds.getNodeDescriptors

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,130 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
 	return err
 }
 
+func (ds *DistSender) sendAttempt(desc *proto.RangeDescriptor, call client.Call) (retry.Status, error) {
```
adding a comment here.

tamird commented on kv/dist_sender.go:
```diff
@@ -375,18 +379,19 @@ func (ds *DistSender) optimizeReplicaOrder(replicas replicaSlice) rpc.OrderingPo
-		log.Infof("unable to determine this node's attributes for replica " +
-			"selection; node is most likely bootstrapping")
+
+	if nodeDesc, err := ds.gossip.GetInfo(gossip.MakeNodeIDKey(ownNodeID)); ownNodeID > 0 && err == nil {
```
`ownNodeID` can be checked before going to gossip?

tschottdorf replied:
done

tamird commented on kv/dist_sender.go:
```diff
@@ -375,18 +379,19 @@ func (ds *DistSender) optimizeReplicaOrder(replicas replicaSlice) rpc.OrderingPo
-			"selection; node is most likely bootstrapping")
+
+	if nodeDesc, err := ds.gossip.GetInfo(gossip.MakeNodeIDKey(ownNodeID)); ownNodeID > 0 && err == nil {
+		atomic.StorePointer(&ds.nodeDescriptor, unsafe.Pointer(nodeDesc.(*proto.NodeDescriptor)))
```
extract the type assertion

tamird replied:
don't we want to lock around the gossip work to prevent two goroutines from duplicating that work?

tschottdorf replied:
it's only a local gossip lookup in the first couple of seconds on the node, not worth a lock.

tamird replied:
why use `atomic`, then?

tschottdorf replied:
because otherwise there is an honest data race whereas now we might end up doing work 3 times for a very short period of time.

tamird replied:
I'd like to see a lock here. Maybe we need a bigger discussion on how we handle cases like this in general.

tschottdorf replied:
done (type assertion). I disagree about the lock because this operation is idempotent, but feel free to go for that after this lands. What would really be useful is avoiding this awkward dance around setting the NodeID late, but that's going to be more involved.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+
+	rpcErr := ds.sendRPC(desc.RaftID, replicas, order, args, reply)
+
+	// For an RPC error to occur, we've must've been unable to contact
```
s/we've/we/

tschottdorf replied:
done.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+	// rate limiting here.
+	if rpcErr != nil {
+		if log.V(1) {
+			log.Warningf("RPC error; expiring cached descriptor")
```
should we log `desc` here?

tschottdorf replied:
I think I'll remove this log statement. In 99.9% of cases the fact that an eviction happens is irrelevant here, and the RPC error gets logged below already.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+		ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+	}
+
+	if err := util.FirstError(rpcErr, reply.Header().GoError()); err != nil {
```
why do we need to do type assertions on `rpcErr` if the only failure case is explained and handled immediately above?

tschottdorf replied:
I don't see any type assertions. The rpcErr is the error returned by the RPC call, and we don't want to mix it up with the error contained in a successful RPC call's response.

tamird replied:
right, but your call to `util.FirstError(rpcErr,...)` might return `rpcErr`, so that `err == rpcErr`, and then you do type assertions below.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+			// we've got a stale replica; evict cache.
+			// Next, cache the new leader.
+			if newLeader != nil {
+				if i := replicas.FindReplica(newLeader.StoreID); i == -1 {
```
`if replicas.FindReplica(newLeader.StoreID) == -1`?

tschottdorf replied:
done.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+			// Next, cache the new leader.
+			if newLeader != nil {
+				if i := replicas.FindReplica(newLeader.StoreID); i == -1 {
+					ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
```
does this deserve a V(1) log?

tschottdorf replied:
yes. I'll add one.

tschottdorf replied:
done. Also updated the code to check the descriptor, not the replica set (the latter excludes peers whose address wasn't found in gossip, so it doesn't tell us anything about the descriptor).

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+
+// getDescriptors takes a call and looks up the corresponding range descriptors
+// associated to it. First, the range descriptor for the corresponding StartKey
+// from call.Args is returned; second, if the EndKey exceeds that of the that
```
s/the EndKey exceeds that of the that descriptor/the EndKey from call.Args exceeds that of the returned range descriptor/

tschottdorf replied:
done.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+	var descNext *proto.RangeDescriptor
+	// If the request accesses keys beyond the end of this range,
+	// get the descriptor of the adjacent range to address next.
+	if desc.EndKey.Less(call.Args.Header().EndKey) {
```
I think we need the adjacent range if these are equal, too, since `desc.EndKey` is exclusive

tschottdorf replied:
Range ops are also exclusive, so this looks right to me.

tamird replied:
Can you point me to where that's documented?

tschottdorf replied:
I'm adding a comment to `proto/api.proto`.

tamird replied:
LG, thanks.

spencerkimball replied:
yes this is correct.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+		// This next lookup is likely for free since we've read the
+		// previous descriptor and range lookups use cache
+		// prefetching.
+		var err error
```
`err` is already declared in this scope, no?

tschottdorf replied:
done.

tamird commented on kv/dist_sender.go:
```diff
@@ -491,154 +606,63 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
-	options := lookupOptions{}
-	if pushArgs, ok := args.(*proto.InternalPushTxnRequest); ok {
-		options.ignoreIntents = pushArgs.RangeLookup
+	// If this is a bounded request, we will change its bound as we receive
```
can this live right next to the described bound-changing?

tschottdorf replied:
The bound-changing may happen repeatedly. I tried, but it looked more awkward than this.

tamird commented on kv/dist_sender.go:
```diff
@@ -491,154 +606,63 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
-		// "Untruncate" EndKey to original. We do this even on error in
-		// case the caller will retry using the same args.
+		// "Untruncate" EndKey to original.
 		args.Header().EndKey = endKey
```
This is superfluous if `descNext` above is nil - perhaps we could run an anonymous function in the `for {` and use `defer` to do this only when needed?

tschottdorf replied:
good idea, done. This even fixes a bug in the original code, I think (truncate, call fails, restart - descNext will be nil though the request spanned ranges). Might be worth a test.

tamird commented on kv/dist_sender.go:
```diff
@@ -647,42 +671,71 @@ func (ds *DistSender) Send(_ context.Context, call client.Call) {
+			// This was the second or later call in a multi-range request.
+			// Combine the new response with the existing one.
+			_, ok := finalReply.(proto.Combinable)
+			if !ok {
```
nit: inline?

tamird replied:
actually no, this type assertion is repeated below - let's just do it once

tschottdorf replied:
That's what I had originally, but thought it looked weird to have

```
cFinalReply, ok := ...
if !ok { ... }
cFinalReply.Combine(curReply)
```

because it obfuscates that it alters `finalReply`.

tamird replied:
Could go the other way:
```go
cFinalReply, ok := finalReply.(proto.Combinable); ok {
  cFinalReply.Combine(curReply)
  continue
}
call.Reply.Header().SetGoError(util.Errorf("multi-range request with non-combinable response type"))
return
```

tschottdorf replied:
That's not equivalent since that code isn't at the end of the surrounding for loop (and even if it were, that seems prone to future errors). I'll stick with the current version.

tamird replied:
ah, right you are. Simple `if - else` would be appropriate, then

tbg committed:
break down DistSender.Send

- remove the mechanism for removing a cached leader upon
  caching the same leader twice in a row: silly idea, obviously
  concurrent request going to the wrong leader will return
  almost simultaneously and expire the freshly updated cache.
- the leaderCache now only deals with proto.Replica{}, not pointers.
- pretty complete refactoring of DistSender.Send to get more manageable
  pieces and less error handling spagetthi.

tbg committed:
add comment about EndKey in multi-range ops

tamird commented on kv/dist_sender.go:
```diff
@@ -491,198 +609,138 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
 	// required, set the timestamp using the local clock.
 	if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Timestamp.Equal(proto.ZeroTimestamp) {
+		// Make sure that after the call, args hasn't changed.
+		defer func() {
```
kind of a nit:
```go
defer func(timestamp proto.Timestamp) {
  args.Header().Timestamp = timestamp
}(args.Header().Timestamp)
```

tschottdorf replied:
I'll keep the explicit zero timestamp, seems more legible.

tamird commented on kv/dist_sender.go:
```diff
@@ -491,198 +609,138 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+		// ScanRequest, check whether enough rows have been retrieved.
+		var origBound int64
+		if args, ok := args.(proto.Bounded); ok && args.GetBound() > 0 {
+			origBound = args.GetBound()
```
nit: since this bound may change repeatedly, this value is not really "orig"

tschottdorf replied:
done.

tamird commented on kv/dist_sender.go:
```diff
@@ -375,18 +379,23 @@ func (ds *DistSender) optimizeReplicaOrder(replicas replicaSlice) rpc.OrderingPo
+		return nil
+	}
+
+	if nodeDesc, err := ds.gossip.GetInfo(gossip.MakeNodeIDKey(ownNodeID)); ownNodeID > 0 && err == nil {
```
s/ownNodeID > 0 &&//

tschottdorf replied:
done.

tamird commented on kv/dist_sender.go:
```diff
@@ -375,18 +379,23 @@ func (ds *DistSender) optimizeReplicaOrder(replicas replicaSlice) rpc.OrderingPo
-		log.Infof("unable to determine this node's attributes for replica " +
-			"selection; node is most likely bootstrapping")
+	if ownNodeID == 0 {
+		return nil
```
probably still want the log below to fire in this case

tschottdorf replied:
done.

tschottdorf commented:
@spencerkimball assigned you to take a look since a lot of things have shifted and subtle things may have happened.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -491,198 +608,138 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+			// sendAttempt below may clear them on certain errors, so we
+			// refresh (likely from the cache) on every retry.
+			desc, descNext, descErr = ds.getDescriptors(call)
+			// If getDescriptors fails, we don't fish for retryable errors.
```
@spencerkimball I think those were retried before, but it seems that any range lookups that could reasonably fail would call Send() themselves and then retry within, correct? Getting something retryable back here would correspond to RPC errors or client retry errors from the lookup (which shouldn't occur).

spencerkimball replied:
I agree. We shouldn't need to retry here.

tamird commented on kv/dist_sender.go:
```diff
@@ -480,6 +472,131 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+		ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+	}
+
+	if err := util.FirstError(rpcErr, reply.Header().GoError()); err != nil {
```
there was a discussion here that github decided to kill:
> @tamird: why do we need to do type assertions on rpcErr if the only failure case is explained and handled immediately above?
> @tschottdorf: I don't see any type assertions. The rpcErr is the error returned by the RPC call, and we don't want to mix it up with the error contained in a successful RPC call's response.
> @tamird: right, but your call to util.FirstError(rpcErr,...) might return rpcErr, so that err == rpcErr, and then you do type assertions below.

tschottdorf replied:
Ok, I see what you mean. We just want to treat an RPC error like any other for simplicity (see the `default` case in the select below).

tamird replied:
Can you achieve this by adding `rpc.SendError` to the first `case`?

spencerkimball commented on kv/dist_sender.go:
```diff
@@ -491,198 +608,138 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
-	if pushArgs, ok := args.(*proto.InternalPushTxnRequest); ok {
-		options.ignoreIntents = pushArgs.RangeLookup
+	// If this is a bounded request, we will change its bound as we receive
+	// replies. this undoes that up when we return.
```
s/that up/that/

tschottdorf replied:
done.

spencerkimball commented on kv/dist_sender.go:
```diff
@@ -480,6 +472,131 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
 }
 
+// sendAttempt is invoked by Send and handles retry logic and cache eviction
+// for a call sent to a single range.
```
Why not mention the return values in comment? In particular, the retry.Status.

    // This method returns a retry status, which is Break on success and either
    // Break, Continue or Reset depending on error condition. This method is
    // expected to be invoked from within a backoff / retry loop to retry the send
    // repeatedly (e.g. to continue processing after a critical node becomes
    // available after downtime or the range descriptor is refreshed via lookup).

tschottdorf replied:
done.

spencerkimball commented:
LGTM
